trieste/models/gpflow/models.py

```diff
@@ -211,6 +232,183 @@ def evaluate_loss_of_model_parameters() -> tf.Tensor:
+        )
+
+        return mean_new, var_new
+
+    def conditional_predict_joint(
+        self, query_points: TensorType, additional_data: Dataset
+    ) -> tuple[TensorType, TensorType]:
+        """
+        Predicts the joint GP distribution at query_points conditioned on both the model
+        and some additional data, using exact formula.
+
+        :param query_points: Set of query points with shape [M, D]
+        :param additional_data: Dataset with query_points with shape [..., N, D] and observations
+                 with shape [..., N, L]
+        :return: mean_new: predictive variance at query_points, with shape [..., M, L],
+                 and cov_new: predictive covariance between query_points, with shape [..., L, M, M]
+        """
+
+        tf.debugging.assert_shapes(
+            [
+                (additional_data.query_points, [..., "N", "D"]),
+                (additional_data.observations, [..., "N", "L"]),
+                (query_points, ["M", "D"]),
+            ],
+            message="additional_data must have query_points with shape [..., N, D]"
+            " and observations with shape [..., N, L], and query_points "
+            "should have shape [M, D]",
+        )
+
+        leading_dims = tf.shape(additional_data.query_points)[:-2]  # [...]
+        new_shape = tf.concat([leading_dims, tf.shape(query_points)], axis=0)  # [..., M, D]
+        query_points_r = tf.broadcast_to(query_points, new_shape)  # [..., M, D]
+        points = tf.concat([additional_data.query_points, query_points_r], axis=-2)  # [..., N+M, D]
+
+        mean, cov = self.model.predict_f(points, full_cov=True)  # [..., N+M, L], [..., L, N+M, N+M]
+
+        N = tf.shape(additional_data.query_points)[-2]
+
+        mean_old = mean[..., :N, :]  # [..., N, L]
+        m_new = mean[..., N:, :]  # [..., M, L]
+
+        cov_old = cov[..., :N, :N]  # [..., L, N, N]
+        c_new = cov[..., N:, N:]  # [..., L, M, M]
+        cov_cross = cov[..., :N, N:]  # [..., L, N, M]
+
+        cov_shape = tf.shape(cov_old)
+        noise = self.model.likelihood.variance * tf.eye(
+            cov_shape[-2], batch_shape=cov_shape[:-2], dtype=cov_old.dtype
+        )
+        L_old = tf.linalg.cholesky(cov_old + noise)  # [..., L, N, N]
+        A = tf.linalg.triangular_solve(L_old, cov_cross, lower=True)  # [..., L, N, M]
+        cov_new = c_new - tf.matmul(A, A, transpose_a=True)  # [..., L, M, M]
+
+        mean_old_diff = additional_data.observations - mean_old  # [..., N, L]
+        mean_old_diff = leading_transpose(mean_old_diff, [..., -1, -2])[..., None]  # [..., L, N, 1]
+        AM = tf.linalg.triangular_solve(L_old, mean_old_diff)  # [..., L, N, 1]
+        mean_new = m_new + leading_transpose(
+            (tf.matmul(A, AM, transpose_a=True)[..., 0]), [..., -1, -2]
+        )  # [..., M, L]
+
+        tf.debugging.assert_shapes(
+            [
+                (additional_data.query_points, [..., "N", "D"]),
```

Comment:
no need to assert again additional_data and query_points, it was done above, right?